File: ts/src/sql/query.ts
export type Primitive = string | number | boolean | null | Date;

export class SqlQuery {
  private tableName = "";
  private selectedFields: string[] = ["*"];
  private filters: string[] = [];
  private sortClause = "";
  private rowLimit?: number;

  static from(table: string): SqlQuery {
    const query = new SqlQuery();
    query.tableName = table;
    return query;
  }

  select(fields: string[]): this {
    this.selectedFields = fields.length > 0 ? fields : ["*"];
    return this;
  }

  where(field: string, value: Primitive): this {
    this.filters.push(`${field} = ${this.render(value)}`);
    return this;
  }

  whereRaw(clause: string): this {
    this.filters.push(clause);
    return this;
  }

  orderBy(field: string, direction: "asc" | "desc" = "asc"): this {
    this.sortClause = ` ORDER BY ${field} ${direction.toUpperCase()}`;
    return this;
  }

  limit(count: number): this {
    this.rowLimit = count;
    return this;
  }

  toString(): string {
    const where = this.filters.length > 0 ? ` WHERE ${this.filters.join(" AND ")}` : "";
    const limit = this.rowLimit === undefined ? "" : ` LIMIT ${this.rowLimit}`;
    return `SELECT ${this.selectedFields.join(", ")} FROM ${this.tableName}${where}${this.sortClause}${limit}`;
  }

  private render(value: Primitive): string {
    if (value === null) return "NULL";
    if (value instanceof Date) return `'${value.toISOString()}'`;
    if (typeof value === "number") return String(value);
    if (typeof value === "boolean") return value ? "TRUE" : "FALSE";
    return `'${value}'`;
  }
}

export function buildInsert(table: string, record: Record<string, Primitive>): string {
  const fields = Object.keys(record);
  const values = fields.map((field) => renderValue(record[field]));
  return `INSERT INTO ${table} (${fields.join(", ")}) VALUES (${values.join(", ")})`;
}

function renderValue(value: Primitive): string {
  if (value === null) return "NULL";
  if (value instanceof Date) return `'${value.toISOString()}'`;
  if (typeof value === "number") return String(value);
  if (typeof value === "boolean") return value ? "TRUE" : "FALSE";
  return `'${value}'`;
}

File: python/robobites_shared/query.py
from datetime import date, datetime
from typing import Any

Primitive = str | int | float | bool | None | date | datetime


class SqlQuery:
    def __init__(self, table: str) -> None:
        self.table = table
        self.fields: list[str] = ["*"]
        self.filters: list[str] = []
        self.sort_clause = ""
        self.row_limit: int | None = None

    @classmethod
    def from_table(cls, table: str) -> "SqlQuery":
        return cls(table)

    def select(self, fields: list[str]) -> "SqlQuery":
        self.fields = fields or ["*"]
        return self

    def where(self, field: str, value: Primitive) -> "SqlQuery":
        self.filters.append(f"{field} = {self._render(value)}")
        return self

    def where_raw(self, clause: str) -> "SqlQuery":
        self.filters.append(clause)
        return self

    def order_by(self, field: str, direction: str = "asc") -> "SqlQuery":
        self.sort_clause = f" ORDER BY {field} {direction.upper()}"
        return self

    def limit(self, count: int) -> "SqlQuery":
        self.row_limit = count
        return self

    def to_sql(self) -> str:
        where = f" WHERE {' AND '.join(self.filters)}" if self.filters else ""
        limit = "" if self.row_limit is None else f" LIMIT {self.row_limit}"
        return f"SELECT {', '.join(self.fields)} FROM {self.table}{where}{self.sort_clause}{limit}"

    def _render(self, value: Primitive) -> str:
        if value is None:
            return "NULL"
        if isinstance(value, bool):
            return "TRUE" if value else "FALSE"
        if isinstance(value, (int, float)):
            return str(value)
        if isinstance(value, (date, datetime)):
            return f"'{value.isoformat()}'"
        return f"'{value}'"


def build_insert(table: str, record: dict[str, Primitive]) -> str:
    fields = list(record.keys())
    values = [_render_value(record[field]) for field in fields]
    return f"INSERT INTO {table} ({', '.join(fields)}) VALUES ({', '.join(values)})"


def _render_value(value: Primitive) -> str:
    if value is None:
        return "NULL"
    if isinstance(value, bool):
        return "TRUE" if value else "FALSE"
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, (date, datetime)):
        return f"'{value.isoformat()}'"
    return f"'{value}'"
